sdk/lib/rtl/rangelist.c

```diff
@@ -13,546 +13,1556 @@
-    PLIST_ENTRY Next;
+    RTL_RANGE_LIST_ITERATOR Iterator;
+    PRTL_RANGE RtlRange;
+    NTSTATUS Status;
 
-    RangeList = CONTAINING_RECORD(Iterator->RangeListHead, RTL_RANGE_LIST, ListHead);
-    if (Iterator->Stamp != RangeList->Stamp)
-        return STATUS_INVALID_PARAMETER;
+    PAGED_CODE_RTL();
+    DPRINT("RtlIsRangeAvailable: [%X] %p, [%I64X-%I64X], %X\n", Flags, RangeList, Start, End, AttributeAvailableMask);
 
-    if (Iterator->Current == NULL)
-    {
-        *Range = NULL;
-        return STATUS_NO_MORE_ENTRIES;
-    }
+    ASSERT(RangeList);
+    ASSERT(Available);
 
-    if (MoveForwards)
-    {
-        Next = ((PRTL_RANGE_ENTRY)Iterator->Current)->Entry.Flink;
-    }
-    else
+    Status = RtlGetFirstRange(RangeList, &Iterator, &RtlRange);
+
+    if (Status == STATUS_NO_MORE_ENTRIES)
     {
-        Next = ((PRTL_RANGE_ENTRY)Iterator->Current)->Entry.Blink;
+        *Available = TRUE;
+        return STATUS_SUCCESS;
     }
 
-    if (Next == Iterator->RangeListHead)
+    if (!NT_SUCCESS(Status))
     {
-        Iterator->Current = NULL;
-        *Range = NULL;
-        return STATUS_NO_MORE_ENTRIES;
+        DPRINT1("RtlIsRangeAvailable: Status %X\n", Status);
+        return Status;
     }
 
-    Iterator->Current = Next;
-    *Range = &((PRTL_RANGE_ENTRY)Next)->Range;
-
+    *Available = RtlpIsRangeAvailable(&Iterator,
+                                      Start,
+                                      End,
+                                      AttributeAvailableMask,
+                                      ((Flags & RTL_RANGE_LIST_SHARED_OK) != 0),
+                                      ((Flags & RTL_RANGE_LIST_NULL_CONFLICT_OK) != 0),
+                                      TRUE,
+                                      Context,
+                                      Callback);
     return STATUS_SUCCESS;
 }
 
-
-/**********************************************************************
- * NAME							EXPORTED
- * 	RtlInitializeRangeList
- *
- * DESCRIPTION
- *	Initializes a range list.
+/**
+ * @name RtlInitializeRangeList
  *
- * ARGUMENTS
- *	RangeList	Pointer to a user supplied range list.
+ * Initializes a range list.
  *
- * RETURN VALUE
- *	None
+ * @param RangeList
+ * Pointer to a user supplied range list.
  *
- * @implemented
+ * @return
+ * None.
  */
+CODE_SEG("PAGE")
+NTSYSAPI
 VOID
 NTAPI
-RtlInitializeRangeList(IN OUT PRTL_RANGE_LIST RangeList)
+RtlInitializeRangeList(
+    _Inout_ PRTL_RANGE_LIST RangeList)
 {
+    PAGED_CODE_RTL();
+    DPRINT("RtlInitializeRangeList: RangeList %p\n", RangeList);
+
+    ASSERT(RangeList);
+
     InitializeListHead(&RangeList->ListHead);
     RangeList->Flags = 0;
     RangeList->Count = 0;
     RangeList->Stamp = 0;
 }
 
+// !!! FIXME ==> PRTLP_RANGE_LIST_ENTRY
+
+typedef struct _RTL_RANGE_ENTRY {
+    RTL_RANGE   Range;
+    USHORT      PrivateFlags;
+    LIST_ENTRY  Entry;
+    UCHAR       _PADDING0_[0x4];
+} RTL_RANGE_ENTRY, *PRTL_RANGE_ENTRY;
```

Comment:
This needs to be cleared up for the final version of the PR

Reply:
At first it seemed to me the "final version of ROS" :)

RTL_RANGE_ENTRY use only in the RtlInvertRangeList() and RtlMergeRangeLists and if they are reworked then this structure will clear itself.